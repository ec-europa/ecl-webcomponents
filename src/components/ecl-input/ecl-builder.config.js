const path = require('path');
const pkg = require('./package.json');

const nodeModules = path.resolve(__dirname, '../../../node_modules');
const outputFolder = './build';

// SCSS includePaths
const includePaths = [nodeModules];

const banner = `${pkg.name} - ${pkg.version} Built on ${new Date().toISOString()}`;

module.exports = {
  scripts: [
    {
      entry: path.resolve(nodeModules, '@ecl/vanilla-component-file-upload/file-upload.js'),
      dest: path.resolve(outputFolder, 'scripts/ecl-file-upload-vanilla.js'),
      options: {
        banner,
        moduleName: 'ECL',
        includePaths,
        sourceMap: false,
      },
    },
    {
      entry: path.resolve(nodeModules, '@ecl/vanilla-component-range/range.js'),
      dest: path.resolve(outputFolder, 'scripts/ecl-range-vanilla.js'),
      options: {
        banner,
        moduleName: 'ECL',
        includePaths,
        sourceMap: false,
      },
    },
  ],
  styles: [
    {
      entry: path.resolve(__dirname, 'src/styles/ecl-input-ec.scss'),
      dest: path.resolve(outputFolder, 'styles/ecl-input-ec.css'),
      options: {
        banner,
        includePaths,
        sourceMap: false,
      },
    },
    {
      entry: path.resolve(__dirname, 'src/styles/ecl-input-eu.scss'),
      dest: path.resolve(outputFolder, 'styles/ecl-input-eu.css'),
      options: {
        banner,
        includePaths,
        sourceMap: false,
      },
    },
  ],
};
