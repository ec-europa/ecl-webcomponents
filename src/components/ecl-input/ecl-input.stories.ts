const getArgs = () => {
  return {
    type: 'text',
    width: 'm',
    disabled: false,
    required: true,
    invalid: false,
    inputId: 'input-id',
    name: 'input-name',
    label: 'Item label',
    placeholder: 'Placeholder text',
    defaultValue: '',
  };
};

const getArgTypes = () => {
  return {
    type: {
      type: { name: 'select' },
      options: ['text', 'search', 'radio', 'checkbox'],
      description: 'Input type (text, radio, search, checkbox)',
      table: {
        category: 'Input type',
      },
    },
    width: {
      type: { name: 'select' },
      options: ['s', 'm', 'l'],
      description: 'Width of the input',
      table: {
        category: 'Styles',
      },
    },
    disabled: {
      type: { name: 'boolean' },
      description: 'Disabled input',
      table: {
        category: 'States',
      },
    },
    required: {
      type: { name: 'boolean' },
      description: 'Required input',
      table: {
        category: 'States',
      },
    },
    invalid: {
      type: { name: 'boolean' },
      description: 'Invalid input',
      table: {
        category: 'States',
      },
    },
    inputId: {
      name: 'input id',
      type: { name: 'string' },
      description: 'Input variant',
      table: {
        category: 'Content',
      },
    },
    placeholder: {
      type: { name: 'string' },
      description: 'Placeholder element',
      table: {
        category: 'Content',
      },
    },
    name: {
      type: { name: 'string' },
      description: 'Name attribute',
      table: {
        category: 'Content',
      },
    },
    label: {
      type: { name: 'string' },
      description: 'Checkbox/Radio label',
      table: {
        category: 'Content',
      },
    },
    defaultValue: {
      name: 'value',
      type: { name: 'string' },
      description: 'Default value of the input',
      table: {
        category: 'Content',
      },
    },
  };
};

export default {
  title: 'Components/form/input-field',
};

const Template = args =>
`<ecl-form-group
  label="Label"
  required=${args.required}
  optional-text="(optional)"
  helper-text="This is the input helper text"
  invalid=${args.invalid}
  disabled=${args.disabled}
  invalid-text="This is an error message"
  theme="${args.theme}"
  aria-label-optional="Aria label optional"
  ${args.type === 'checkbox' || args.type === 'radio' ? `tag="${args.type === 'checkbox' || args.type === 'radio' ? 'fieldset' : ''}"` : ''}
  ${args.type === 'checkbox' || args.type === 'radio' ? `label-tag="${args.type === 'checkbox' || args.type === 'radio' ? 'legend' : ''}"` : ''}
  aria-label-required="Aria label required"
>
  <ecl-input
    type="${args.type}"
    theme="${args.theme}"
    input-id="${args.inputId}"
    default-value="${args.defaultValue}"
    label="${(args.type === 'checkbox' || args.type === 'radio') ? args.label : ''}"
    required=${args.required}
    disabled=${args.disabled}
    invalid=${args.invalid}
    width="${args.width}"
    placeholder="${args.placeholder}"
    helper-text="${args.type === 'checkbox' || args.type === 'radio' ? 'This is the input helper text' : '' }"
    name="${args.name}"
  >
  </ecl-input>
</ecl-form-group>`;

export const Input = Template.bind({});
Input.storyName = 'default';
Input.args = getArgs();
Input.argTypes = getArgTypes();
