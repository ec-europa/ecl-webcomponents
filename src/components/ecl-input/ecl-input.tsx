import { Component, h, Prop, Element, getAssetPath } from '@stencil/core';
declare const ECL: any;

@Component({
  tag: 'ecl-input',
  styleUrls: {
    ec: './build/styles/ecl-input-ec.css',
    eu: './build/styles/ecl-input-eu.css',
  },
  shadow: false,
  scoped: true,
  assetsDirs: ['build'],
})
export class EclInput {
  @Element() el: HTMLElement;
  @Prop() theme: string = 'ec';
  @Prop() styleClass: string;
  @Prop() inputClass: string;
  @Prop() eclScript: boolean = false;
  @Prop() withUtils: boolean = false;
  @Prop() disabled: boolean = false;
  @Prop() required: boolean = false;
  @Prop() invalid: boolean = false;
  @Prop() helperId: string;
  @Prop() helperText: string;
  @Prop() placeholder: string;
  @Prop() width: string = 'm';
  @Prop() label: string;
  @Prop() type: string;
  @Prop() inputId: string;
  @Prop() name: string;
  @Prop() defaultValue: string;
  @Prop() max: number;
  @Prop() min: number;
  @Prop() step: number = 1;
  @Prop() valueLabel: string;
  @Prop() buttonChooseLabel: string;
  @Prop() buttonReplaceLabel: string;
  @Prop() multiple: boolean = false;

  getClass(): string {
    const styleClasses = ['ecl-input', this.styleClass];
    if (this.type === 'checkbox') {
      styleClasses.push('ecl-checkbox', `ecl-checkbox--${this.width}`);

      if (this.disabled) {
        styleClasses.push('ecl-checkbox--disabled');
      }
      if (this.invalid) {
        styleClasses.push('ecl-checkbox--invalid');
      }
    }
    if (this.type === 'radio') {
      styleClasses.push('ecl-radio', `ecl-radio--${this.width}`);

      if (this.disabled) {
        styleClasses.push('ecl-radio--disabled');
      }
      if (this.invalid) {
        styleClasses.push('ecl-radio--invalid');
      }
    }

    return styleClasses.join(' ');
  }

  getInputClasses(input): string {
    let inputClasses = [`ecl-${input}__input`];

    if (this.type === 'range') {
      inputClasses = ['ecl-range', `ecl-range--${this.width}`];
    }
    if (this.type === 'file') {
      inputClasses = ['ecl-file-upload'];
    }
    if (this.type === 'text' || this.type === 'search') {
      inputClasses = ['ecl-text-input', `ecl-text-input--${this.width}`];
      if (this.invalid) {
        inputClasses.push('ecl-text-input--invalid');
      }
      if (this.disabled) {
        inputClasses.push('ecl-text-input--disabled');
      }
    }
    if (this.inputClass) {
      inputClasses.push(this.inputClass);
    }

    return inputClasses.join(' ');
  }

  getBoxClasses(input, box): string {
    let boxClasses = [`ecl-${input}__${box}`];
    if (this.disabled) {
      boxClasses.push(`ecl-${input}__${box}--disabled`);
    }
    if (this.invalid) {
      boxClasses.push(`ecl-${input}__${box}--invalid`);
    }

    return boxClasses.join(' ');
  }

  componentWillLoad() {
    if (this.withUtils && !document.querySelector('#ecl-utils-styles')) {
      const style = document.createElement('link');
      style.rel = 'stylesheet';
      style.type = 'text/css';
      style.id = 'ecl-utils-styles';
      style.href = getAssetPath(`./build/styles/ecl-utilities-${this.theme}.css`);
      document.body.appendChild(style);
    }
  }

  componentDidRender() {
    if (this.type === 'range' && this.eclScript) {
      const src = getAssetPath('./build/scripts/ecl-range-vanilla.js');
      if (document.querySelector(`script[src="${src}"]`)) {
        document.querySelector(`script[src="${src}"]`).remove();
      }
      const script = document.createElement('script');
      script.src = src;
      script.onload = () => {
        const formGroup = this.el.closest('.ecl-form-group');
        formGroup.setAttribute('data-ecl-range', 'data-ecl-range');
        const range = new ECL.Range(formGroup);
        range.init();
      };
      document.body.appendChild(script);
    }
    if (this.type === 'file' && this.eclScript) {
      const src = getAssetPath('./build/scripts/ecl-file-upload-vanilla.js');
      if (document.querySelector(`script[src="${src}"]`)) {
        document.querySelector(`script[src="${src}"]`).remove();
      }
      const script = document.createElement('script');
      script.src = src;
      script.onload = () => {
        const fileUpload = new ECL.FileUpload(this.el.firstElementChild);
        fileUpload.init();
      };
      // @ts-ignore
      const observer = new MutationObserver((mutationsList, observer) => {
        for (const mutation of mutationsList) {
          if (mutation.type === 'childList') {
            const nodes = mutation.addedNodes;
            nodes.forEach(node => {
              const htmlNode = node as HTMLElement;
              htmlNode.classList.add(`sc-ecl-input-${this.theme}`);
              const name = htmlNode.querySelector('.ecl-file-upload__item-name');
              if (name) {
                name.classList.add(`sc-ecl-input-${this.theme}`);
              }
              const meta = htmlNode.querySelector('.ecl-file-upload__item-meta');
              if (meta) {
                meta.classList.add(`sc-ecl-input-${this.theme}`);
              }
            });
          }
        }
      });

      observer.observe(this.el.querySelector('.ecl-file-upload__list'), {
        attributes: true,
        childList: true,
        subtree: true,
      });
      document.body.appendChild(script);
    }
  }

  render() {
    const wrapperAttrs = {};
    if (this.type === 'range') {
      wrapperAttrs['data-ecl-range'] = true;
    }
    if (this.type === 'file') {
      wrapperAttrs['data-ecl-file-upload-group'] = true
    }
    const attributes = {
      class: this.getInputClasses(this.type),
      type: this.type,
      id: this.inputId,
      name: this.name,
      required: this.required,
      disabled: this.disabled,
      value: this.defaultValue,
      placeholder: this.placeholder,
    };

    if (this.type === 'file') {
      attributes['data-ecl-file-upload-input'] = true;
      attributes['multiple'] = this.multiple;
    }

    if (this.type === 'range') {
      attributes['data-ecl-range-input'] = true;
      attributes['min'] = this.min;
      attributes['max'] = this.max;
      attributes['step'] = this.step;
    }

    return (
      <div 
        class={this.getClass()}
        {...wrapperAttrs}
      >
        <input {...attributes} />
      { this.type === 'checkbox' ?
        <label
          class={this.getBoxClasses('checkbox', 'label')}
          htmlFor={this.inputId}
        >
          <span class={this.getBoxClasses('checkbox', 'box')}>
            <ecl-icon
              style-class={`ecl-checkbox__icon sc-ecl-input-${this.theme}`}
              icon="check"
              size="m"
              ></ecl-icon>
          </span>
          <span class="ecl-checkbox__text">{this.label}</span>
        </label>
        : ''
      }
      { this.type === 'checkbox' && this.helperText ?
        <div
          class="ecl-checkbox__help"
          id={this.helperId}
        >
          {this.helperText}
        </div>
        : ''
      }
      { this.type === 'radio' ?
        <label 
          class={this.getBoxClasses('radio', 'label')}
          htmlFor={this.inputId}
        >
          <span class={this.getBoxClasses('radio', 'box')}>
            <span class="ecl-radio__box-inner"></span>
          </span>
          <span class="ecl-radio__text">{this.label}</span>
        </label>
        : ''
      }
      { this.type === 'radio' && this.helperText ?
        <div 
          class="ecl-radio__help"
          id={this.helperId}
        >
          {this.helperText}
        </div>
        : ''
      }
      { this.type === 'range' ?
        <div class="ecl-range__value">
          {this.valueLabel}
          <span class="ecl-range__value-current" data-ecl-range-value-current> </span>
        </div> : ''
      }
      { this.type === 'file' ?
        <label
          class="ecl-file-upload__button-container"
          htmlFor={this.inputId}
        >
          <span
            class="ecl-file-upload__button ecl-button ecl-button--primary"
            data-ecl-file-upload-button
            data-ecl-file-upload-label-choose={this.buttonChooseLabel}
            data-ecl-file-upload-label-replace={this.buttonReplaceLabel}
          >
            {this.buttonChooseLabel}
          </span>
        </label> : '' 
      }
      { this.type === 'file' ?
        <ul
          class="ecl-file-upload__list"
          data-ecl-file-upload-list
          > 
        </ul> : ''
      }
      </div>
    );
  }
}