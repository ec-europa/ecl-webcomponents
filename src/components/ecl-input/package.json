{
  "name": "@ecl/web-components-input",
  "author": "European Commission",
  "license": "EUPL-1.2",
  "version": "0.5.0",
  "description": "ECL Select",
  "main": "ecl-input.tsx",
  "private": true,
  "scripts": {
    "build:scripts": "cross-env ecl-builder scripts",
    "build:styles": "cross-env ecl-builder styles",
    "build": "run-s clean build:*",
    "clean": "rimraf build",
    "prepublish": "npm run build"
  },
  "devDependencies": {
    "@ecl/builder": "3.8.4",
    "@ecl/vanilla-component-button": "3.8.4",
    "@ecl/vanilla-component-checkbox": "3.8.4",
    "@ecl/vanilla-component-datepicker": "3.8.4",
    "@ecl/vanilla-component-file-upload": "3.8.4",
    "@ecl/vanilla-component-radio": "3.8.4",
    "@ecl/vanilla-component-search-form": "3.8.4",
    "@ecl/vanilla-component-text-input": "3.8.4",
    "cross-env": "7.0.3"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/planctus/ecl-webcomponents.git"
  },
  "bugs": {
    "url": "https://github.com/planctus/ecl-webcomponents/issues"
  },
  "homepage": "https://github.com/planctus/ecl-webcomponents",
  "keywords": [
    "ecl",
    "europa-component-library",
    "design-system",
    "web-components"
  ]
}
