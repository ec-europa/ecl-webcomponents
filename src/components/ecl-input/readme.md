# ecl-input



<!-- Auto Generated Below -->


## Properties

| Property       | Attribute       | Description | Type      | Default     |
| -------------- | --------------- | ----------- | --------- | ----------- |
| `defaultValue` | `default-value` |             | `string`  | `undefined` |
| `disabled`     | `disabled`      |             | `boolean` | `false`     |
| `eclScript`    | `ecl-script`    |             | `boolean` | `false`     |
| `helperId`     | `helper-id`     |             | `string`  | `undefined` |
| `helperText`   | `helper-text`   |             | `string`  | `undefined` |
| `inputId`      | `input-id`      |             | `string`  | `undefined` |
| `invalid`      | `invalid`       |             | `boolean` | `false`     |
| `label`        | `label`         |             | `string`  | `undefined` |
| `max`          | `max`           |             | `number`  | `undefined` |
| `min`          | `min`           |             | `number`  | `undefined` |
| `name`         | `name`          |             | `string`  | `undefined` |
| `placeholder`  | `placeholder`   |             | `string`  | `undefined` |
| `required`     | `required`      |             | `boolean` | `false`     |
| `styleClass`   | `style-class`   |             | `string`  | `undefined` |
| `theme`        | `theme`         |             | `string`  | `'ec'`      |
| `type`         | `type`          |             | `string`  | `undefined` |
| `width`        | `width`         |             | `string`  | `'m'`       |


## Dependencies

### Depends on

- [ecl-icon](../ecl-icon)

### Graph
```mermaid
graph TD;
  ecl-input --> ecl-icon
  style ecl-input fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
