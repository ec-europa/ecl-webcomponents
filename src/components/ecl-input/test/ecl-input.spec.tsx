import { newSpecPage } from '@stencil/core/testing';
import { EclInput } from '../ecl-input';

describe('ecl-input', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [EclInput],
      html: `<ecl-input></ecl-input>`,
    });
    expect(page.root).toEqualHtml(`
      <ecl-input>
        <mock:shadow-root>
          <slot></slot>
        </mock:shadow-root>
      </ecl-input>
    `);
  });
});
