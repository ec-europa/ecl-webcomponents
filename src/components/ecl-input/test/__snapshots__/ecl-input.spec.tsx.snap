// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ecl-input renders 1`] = `
<ecl-form-group helper-text="This is the form group helper text." label="Label" optional-text="(optional)" required="true" theme="ec">
  <!---->
  <div class="ecl-form-group">
    <label class="ecl-form-label" for="input-id" id="input-id-label">
      Label
      <span class="ecl-form-label__required">
        *
      </span>
    </label>
    <div class="ecl-help-block" id="input-id-helper">
      This is the form group helper text.
    </div>
    <ecl-input input-id="input-id" placeholder="placeholder" required="true" theme="ec" type="text" width="l">
      <div class="ecl-input">
        <input class="ecl-text-input ecl-text-input--l" id="input-id" placeholder="placeholder" required="" type="text">
      </div>
    </ecl-input>
  </div>
</ecl-form-group>
`;
